jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassFinder.java

```diff
@@ -51,13 +55,80 @@ public Class findClass(@Nullable String name) {
     }
   }
 
+  /**
+   * If {@code mainClass} is {@code null}, tries to infer main class in this order:
+   *
+   * <ul>
+   *   <li>1. Looks in a {@code jar} plugin provided by {@code projectProperties} ({@code
+   *       maven-jar-plugin} for maven or {@code jar} task for gradle).
+   *   <li>2. Searches for a class defined with a main method.
+   * </ul>
+   *
+   * <p>Warns if main class is not valid, or throws an error if no valid main class is not found.
+   */
+  public static String resolveMainClass(
+      @Nullable String mainClass, ProjectProperties projectProperties) {
+    BuildLogger logger = projectProperties.getLogger();
+    if (mainClass == null) {
+      logger.info(
+          "Searching for main class... Add a 'mainClass' configuration to '"
+              + projectProperties.getPluginName()
```

Comment:
Oh, `HelpfulSuggestions` was actually intended to be used for things like this. The `HelpfulSuggestionsProvider` would construct with different values for maven vs gradle, and then a method would construct the message to be used in the core code like here. Same goes for the message below about the jar plugin vs task.